chord/chord.go

```diff
@@ -92,99 +91,124 @@ func NewChordNode(listener net.Listener) (*ChordNode, error) {
+	}
+}
+
+func (chordNode *ChordNode) setPredecessor(nodePtr *Node) {
+	chordNode.predecessor.Lock()
+	defer chordNode.predecessor.Unlock()
+	chordNode.predecessor.nodePtr = nodePtr
+}
+
+// returns Node, doesExist
+func (chordNode *ChordNode) getPredecessor() (Node, bool) {
+	chordNode.predecessor.RLock()
+	defer chordNode.predecessor.RUnlock()
+	if chordNode.predecessor.nodePtr == nil {
+		return Node{}, false
+	} else {
+		return *chordNode.predecessor.nodePtr, true
+	}
+}
+
 // Create creates a Chord ring
 func (chordNode *ChordNode) Create() {
 	// create a new Chord ring.
 	// n.create()
 	//   predecessor = nil;
-	//	 successor = [n];
-	chordNode.predecessor.Lock()
-	chordNode.predecessor.node = nil
-	chordNode.predecessor.Unlock()
-
-	chordNode.successorsList.Lock()
-	chordNode.successorsList.list = []*Node{&chordNode.node}
-	chordNode.successorsList.Unlock()
+	//   successor = n;
+	chordNode.setPredecessor(nil)
+	chordNode.setSuccessor(&chordNode.node)
 }
 
 // Join lets the ChordNode to join a Chord ring containing node n0.
 // Algorithm:
 // n.join(n0)
-//   predecessor = nil;
-//   succ := n0.find_successor(n);
-//	 fingerTable[0] = succ
-//   succList := succ.getSuccessorsList();
-//   successorsList = succ :: succList[:len(succList) - 1]
+//	 predecessor = nil;
+//	 successor = n0.find_successor(n);
 func (chordNode *ChordNode) Join(n0 Node) error {
 	var err error
 
-	chordNode.predecessor.Lock()
-	chordNode.predecessor.node = nil
-	chordNode.predecessor.Unlock()
+	chordNode.setPredecessor(nil)
 
-	n0Stub, err := chordNode.getStubFor(ipAddr(n0.Ip))
+	succ, err := chordNode.stubFindSuccessor(ipAddr(n0.Ip), context.Background(), &ID{Id: chordNode.node.Id})
+	fmt.Printf("Im joining succ: %v", succ)
 	if err != nil {
 		return err
 	}
-	succ, err := n0Stub.FindSuccessor(context.Background(), &ID{Id: chordNode.node.Id})
-	if err != nil {
-		return err
-	}
-
-	chordNode.fingerTable.Lock()
-	chordNode.fingerTable.table = []Node{*succ}
-	chordNode.fingerTable.Unlock()
-
-	succStub, err := chordNode.getStubFor(ipAddr(succ.Ip))
-	if err != nil {
-		return err
-	}
-	nodesPtr, err := succStub.GetSuccessorsList(context.Background(), &empty.Empty{})
-	switch {
-	case err != nil:
-		return err
-	case nodesPtr == nil:
-		return fmt.Errorf("successors list from %s is nil", succ.Ip)
-	}
-	succList := nodesPtr.NodeArray
 
-	chordNode.successorsList.Lock()
-	defer chordNode.successorsList.Unlock()
-	chordNode.successorsList.list = append([]*Node{succ}, succList[:len(succList)-1]...)
+	chordNode.setSuccessor(succ)
 
 	return nil
 }
 
+func (chordNode *ChordNode) ClosestPrecedingNode(nodeID nodeID) Node {
+	id := string(nodeID)
+	n := chordNode.node.Id
+	chordNode.fingerTable.RLock()
+	defer chordNode.fingerTable.RUnlock()
+	fingerTable := chordNode.fingerTable.table
+	for i := len(chordNode.fingerTable.table) - 1; i >= 0; i-- {
+		finger := fingerTable[i]
+		if finger == nil {
+			continue
+		}
+		if n < finger.Id && finger.Id < id {
```

Comment:
What happens if `n is greater than id` ?
Let's say n == 219 and id == 33.   
For a finger.Id  == 250 this condition will never be true.

Reply:
then n itself is returned

Reply:
I think finger.Id should be returned